File: benchmarks/julia/b12_ga.jl
# Genetic Algorithm minimizing the Rosenbrock function —
# Computational Intelligence benchmark (MICAI). Tournament selection, uniform
# crossover, uniform mutation. Rosenbrock + uniform mutation use only +,-,*,/,
# so the search is bit-exact across languages. Every random decision draws from
# the shared 64-bit LCG (seed 42) in identical order in all six languages.
# Checksum = best fitness found + sum of the best individual's genes (6 dp).
using Printf

const D = 30
const P = 5000
const G = 1200
const T = 3
const MUT_RATE = 0.1
const MUT_STEP = 0.1

mutable struct SimpleRNG
    state::UInt64
end

function next_double!(rng::SimpleRNG)::Float64
    rng.state = rng.state * UInt64(6364136223846793005) + UInt64(1442695040888963407)
    return Float64(rng.state >> 33) / Float64(UInt64(1) << 31)
end

# column-major: pop[d, p] is contiguous per individual.
function rosenbrock(pop::Matrix{Float64}, p::Int)::Float64
    f = 0.0
    for i in 1:(D - 1)
        a = pop[i + 1, p] - pop[i, p] * pop[i, p]
        b = 1.0 - pop[i, p]
        f += 100.0 * a * a + b * b
    end
    return f
end

function main()
    rng = SimpleRNG(UInt64(42))
    pop = zeros(Float64, D, P)
    newpop = zeros(Float64, D, P)
    fitness = zeros(Float64, P)
    best_genes = zeros(Float64, D)

    for p in 1:P
        for d in 1:D
            pop[d, p] = (next_double!(rng) * 2.0 - 1.0) * 5.0
        end
    end

    best_fit = 1e300
    for g in 1:G
        for p in 1:P
            f = rosenbrock(pop, p)
            fitness[p] = f
            if f < best_fit
                best_fit = f
                for d in 1:D
                    best_genes[d] = pop[d, p]
                end
            end
        end
        for d in 1:D
            newpop[d, 1] = best_genes[d]
        end
        for i in 2:P
            a = Int(floor(next_double!(rng) * P)) + 1
            for t in 2:T
                idx = Int(floor(next_double!(rng) * P)) + 1
                if fitness[idx] < fitness[a]
                    a = idx
                end
            end
            b = Int(floor(next_double!(rng) * P)) + 1
            for t in 2:T
                idx = Int(floor(next_double!(rng) * P)) + 1
                if fitness[idx] < fitness[b]
                    b = idx
                end
            end
            for d in 1:D
                newpop[d, i] = next_double!(rng) < 0.5 ? pop[d, a] : pop[d, b]
            end
            for d in 1:D
                if next_double!(rng) < MUT_RATE
                    newpop[d, i] += (next_double!(rng) * 2.0 - 1.0) * MUT_STEP
                end
            end
        end
        for p in 1:P
            for d in 1:D
                pop[d, p] = newpop[d, p]
            end
        end
    end

    gsum = 0.0
    for d in 1:D
        gsum += best_genes[d]
    end
    @printf("%.6f %.6f\n", best_fit, gsum)
end

main()

File: benchmarks/go/b12_ga.go
// Genetic Algorithm minimizing the Rosenbrock function —
// Computational Intelligence benchmark (MICAI). Tournament selection, uniform
// crossover, uniform mutation. Rosenbrock + uniform mutation use only +,-,*,/,
// so the search is bit-exact across languages. Every random decision draws from
// the shared 64-bit LCG (seed 42) in identical order in all six languages.
// Checksum = best fitness found + sum of the best individual's genes (6 dp).
package main

import "fmt"

const (
	D        = 30
	P        = 5000
	G        = 1200
	T        = 3
	MUT_RATE = 0.1
	MUT_STEP = 0.1
)

type SimpleRNG struct{ state uint64 }

func (r *SimpleRNG) next() float64 {
	r.state = r.state*6364136223846793005 + 1442695040888963407
	return float64(r.state>>33) / float64(uint64(1)<<31)
}

var (
	pop       [P][D]float64
	newpop    [P][D]float64
	fitness   [P]float64
	bestGenes [D]float64
)

func rosenbrock(x *[D]float64) float64 {
	f := 0.0
	for i := 0; i < D-1; i++ {
		a := x[i+1] - x[i]*x[i]
		b := 1.0 - x[i]
		f += 100.0*a*a + b*b
	}
	return f
}

func main() {
	rng := SimpleRNG{42}
	for p := 0; p < P; p++ {
		for d := 0; d < D; d++ {
			pop[p][d] = (rng.next()*2.0 - 1.0) * 5.0
		}
	}

	bestFit := 1e300
	for g := 0; g < G; g++ {
		for p := 0; p < P; p++ {
			f := rosenbrock(&pop[p])
			fitness[p] = f
			if f < bestFit {
				bestFit = f
				for d := 0; d < D; d++ {
					bestGenes[d] = pop[p][d]
				}
			}
		}
		for d := 0; d < D; d++ {
			newpop[0][d] = bestGenes[d]
		}
		for i := 1; i < P; i++ {
			a := int(rng.next() * P)
			for t := 1; t < T; t++ {
				idx := int(rng.next() * P)
				if fitness[idx] < fitness[a] {
					a = idx
				}
			}
			b := int(rng.next() * P)
			for t := 1; t < T; t++ {
				idx := int(rng.next() * P)
				if fitness[idx] < fitness[b] {
					b = idx
				}
			}
			for d := 0; d < D; d++ {
				if rng.next() < 0.5 {
					newpop[i][d] = pop[a][d]
				} else {
					newpop[i][d] = pop[b][d]
				}
			}
			for d := 0; d < D; d++ {
				if rng.next() < MUT_RATE {
					newpop[i][d] += (rng.next()*2.0 - 1.0) * MUT_STEP
				}
			}
		}
		for p := 0; p < P; p++ {
			for d := 0; d < D; d++ {
				pop[p][d] = newpop[p][d]
			}
		}
	}

	gsum := 0.0
	for d := 0; d < D; d++ {
		gsum += bestGenes[d]
	}
	fmt.Printf("%.6f %.6f\n", bestFit, gsum)
}
